### PR TITLE
Add report formatter concept for customization of report output

### PR DESCRIPTION
**Pull request by zanieb**
Adds a `ReportFormatter` trait and a `Reporter::report_with_formatter` method allowing customization of formatting the report without reimplementing reporter internals. Currently, we use `format_external` and `format_terms` but we can refine formatting in the future as needed e.g. all the `explain` methods in the reporter.

Adds a `DefaultStringReportFormatter` which implements the existing format for the `DefaultStringReporter`.

You can see how much easier this is to use in the example diff at https://github.com/zanieb/pubgrub/pull/10/commits/fd30571b6346ca6522067b98f3cb4fe0e082b65f

### CODE REVIEW

**Comment by zanieb** [comment]
I pass the `&impl ReportFormatter` type around instead of adding a `&dyn ReportFormatter` field to the reporter to avoid proliferation of lifetimes. Perhaps a more elegant solution around for someone who wants to fight with the compiler.

**Comment by Eh2406** [comment]
TLDR. I love it!

I had been working toward something like this in https://github.com/pubgrub-rs/pubgrub/tree/reporter_trait, if there are any of those changes that look like good ideas please take them. But your re-factor seems much more user focused and a better place to start.

I would eventually like our reporting infrastructure to support directly outputting to `&mut Write` instead of having to be buffered in strings, but having seen your branch it is a re-factor better done separately.

**Comment by zanieb** [comment]
Good to hear! Here's an upstream pull request https://github.com/pubgrub-rs/pubgrub/pull/158

> I had been working toward something like this in https://github.com/pubgrub-rs/pubgrub/tree/reporter_trait, if there are any of those changes that look like good ideas please take them. But your re-factor seems much more user focused and a better place to start.

Ah interesting. I agree it seems a bit more complicated than we need to get started. I think there are some ideas there we can build on though.

> I would eventually like our reporting infrastructure to support directly outputting to &mut Write instead of having to be buffered in strings, but having seen your branch it is a re-factor better done separately.

Agreed!

**Comment by zanieb** [comment]
Closing in favor of https://github.com/pubgrub-rs/pubgrub/pull/158
